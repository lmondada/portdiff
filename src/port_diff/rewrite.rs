use std::collections::BTreeMap;

use crate::{Port, PortDiff, PortEdge, UniqueVertex};

type V = UniqueVertex;

impl<P: Clone + Ord> PortDiff<V, P> {
    /// Replace the internal edges with `new_edges`.
    ///
    /// All vertices with a changed set of ports are replaced with new vertices.
    ///
    /// Optionally, indicate new boundary ports. The set of new boundary ports
    /// must be of the same length. If a boundary port is `None`, the existing
    /// boundary port is retained.
    pub fn rewrite(
        &self,
        new_edges: &[PortEdge<V, P>],
        new_boundary_ports: Vec<Option<Port<V, P>>>,
    ) -> Result<Self, String> {
        if new_boundary_ports.len() != self.boundary_ports.len() {
            return Err("Mismatching number of boundary ports".to_string());
        }

        // Get port sets to compare
        let new_ports = get_port_lists(new_edges);
        let curr_ports = get_port_lists(self.edges.iter());

        // Create the new vertices where required
        let mut new_vertices = BTreeMap::<V, V>::new();
        for (&curr_v, curr_ports) in curr_ports.iter() {
            let new_ports = new_ports.get(&curr_v);
            if Some(curr_ports) != new_ports {
                let new_v = UniqueVertex::new();
                new_vertices.insert(curr_v, new_v);
            }
        }

        // Lift the vertex name change to ports
        let new_port = |Port { node, port }: &Port<_, P>| {
            let &node = new_vertices.get(node).unwrap_or(node);
            Port {
                node,
                port: port.clone(),
            }
        };

        // Create the new edges
        let new_edges = new_edges.iter().map(|e| PortEdge {
            left: new_port(&e.left),
            right: new_port(&e.right),
        });

        // Create new boundary ports
        let boundary_ports = self
            .boundary_ports
            .iter()
            .zip(&new_boundary_ports)
            .map(|(p, new_p)| new_p.as_ref().unwrap_or(p))
            .map(new_port)
            .collect();

        Ok(Self {
            edges: new_edges.collect(),
            boundary_ports,
            boundary_anc: self.boundary_anc.clone(),
            boundary_desc: self.boundary_desc.clone(),
        })
    }
}

type PortLists<P> = BTreeMap<V, Vec<P>>;

fn get_port_lists<'e, P: Ord + Clone + 'e>(
    edges: impl IntoIterator<Item = &'e PortEdge<V, P>>,
) -> PortLists<P> {
    let mut adjacency = PortLists::new();
    for PortEdge { left, right } in edges {
        for src in [left, right] {
            adjacency
                .entry(src.node)
                .or_default()
                .push(src.port.clone());
        }
    }
    for adj_list in adjacency.values_mut() {
        adj_list.sort_unstable();
    }
    adjacency
}
