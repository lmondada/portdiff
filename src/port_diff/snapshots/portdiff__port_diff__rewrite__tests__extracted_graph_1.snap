---
source: src/port_diff/rewrite.rs
expression: pg.dot_string()
---
digraph {
0 [shape=plain label=<<table border="1"><tr><td align="text" border="0" colspan="3">0</td></tr><tr><td port="out0" align="text" colspan="1" cellpadding="1" >0</td><td port="out1" align="text" colspan="1" cellpadding="1" >1</td><td port="out2" align="text" colspan="1" cellpadding="1" >2</td></tr></table>>]
0:out0 -> 2:in0 [style=""]
0:out1 -> 2:in1 [style=""]
0:out2 -> 2:in2 [style=""]
1 [shape=plain label=<<table border="1"><tr><td port="in0" align="text" colspan="1" cellpadding="1" >0</td><td port="in1" align="text" colspan="1" cellpadding="1" >1</td><td port="in2" align="text" colspan="1" cellpadding="1" >2</td></tr><tr><td align="text" border="0" colspan="3">1</td></tr></table>>]
2 [shape=plain label=<<table border="1"><tr><td port="in0" align="text" colspan="4" cellpadding="1" >0</td><td port="in1" align="text" colspan="4" cellpadding="1" >1</td><td port="in2" align="text" colspan="4" cellpadding="1" >2</td></tr><tr><td align="text" border="0" colspan="12">2</td></tr><tr><td port="out0" align="text" colspan="3" cellpadding="1" >0</td><td port="out1" align="text" colspan="3" cellpadding="1" >1</td><td port="out2" align="text" colspan="3" cellpadding="1" >2</td><td port="out3" align="text" colspan="3" cellpadding="1" >3</td></tr></table>>]
2:out0 -> 4:in0 [style=""]
2:out3 -> 3:in0 [style=""]
3 [shape=plain label=<<table border="1"><tr><td port="in0" align="text" colspan="1" cellpadding="1" >0</td></tr><tr><td align="text" border="0" colspan="1">3</td></tr></table>>]
4 [shape=plain label=<<table border="1"><tr><td port="in0" align="text" colspan="4" cellpadding="1" >0</td></tr><tr><td align="text" border="0" colspan="4">4</td></tr><tr><td port="out0" align="text" colspan="1" cellpadding="1" >0</td><td port="out1" align="text" colspan="1" cellpadding="1" >1</td><td port="out2" align="text" colspan="1" cellpadding="1" >2</td><td port="out3" align="text" colspan="1" cellpadding="1" >3</td></tr></table>>]
4:out0 -> 1:in0 [style=""]
4:out1 -> 1:in1 [style=""]
4:out2 -> 1:in2 [style=""]
4:out3 -> 5:in0 [style=""]
5 [shape=plain label=<<table border="1"><tr><td port="in0" align="text" colspan="1" cellpadding="1" >0</td></tr><tr><td align="text" border="0" colspan="1">5</td></tr></table>>]
}
